File: core/src/ru/myitschool/cubegame/effects/EffectArray.java
package ru.myitschool.cubegame.effects;

import java.util.ArrayList;

/**
 * Created by Voyager on 04.08.2017.
 */
public class EffectArray extends ArrayList<Effect> {

    int id = 0; //TODO reset after exiting of dungeon

    @Override
    public boolean add(Effect effect) {
        add(effect, size());
        return true;
    }

    public void add(Effect effect, int pos) {
        int size = super.size();
        if (effect.isStackable()){
            int stack = effect.getStackSize() - 1;
            int count = 0;
            for (int i = size - 1; i >= 0; i--){
                Effect oldEffet = get(i);
                if (oldEffet.getType() == effect.getType()){
                    count++;
                    if (count >= stack){
                        remove(i);
                    }
                }
            }
        } else {
            for (int i = size - 1; i >= 0; i--) {
                Effect oldEffet = get(i);
                if (oldEffet.getType() == effect.getType()){
                    remove(i);
                }
            }
        }
        if (pos > size()){
            pos = size();
        }
        super.add(pos, effect);
        effect.setId(id++);
    }

    public void removeId(int id){
        for (int i = 0; i < size(); i++) {
            Effect effect = get(i);
            if (effect.getId() == id) {
                remove(i);
                i--;
            }
        }
    }
}

File: core/src/com/wolg_vlad/dcrawler/effects/EffectArray.kt
package com.wolg_vlad.dcrawler.effects

/**
 * Created by Voyager on 04.08.2017.
 */
class EffectArray : ArrayList<Effect>() {
    var id = 0 //TODO reset after exiting of dungeon
    override fun add(effect: Effect): Boolean {
        add(size, effect)
        return true
    }

    override fun add(pos: Int, effect: Effect) {
        var pos = pos
        val initialSize = super.size
        if (effect.stackable) {
            val stack = effect.stackSize - 1
            var count = 0
            for (i in initialSize - 1 downTo 0) {
                val oldEffet = get(i)
                if (oldEffet.id == effect.id) { //todo rework this to remove oldest stack or find another way to work around it
                    count++
                    if (count >= stack) {
                        removeAt(i)
                    }
                }
            }
        } else {
            for (i in initialSize - 1 downTo 0) {
                val oldEffet = get(i)
                if (oldEffet.id == effect.id) {
                    removeAt(i)
                }
            }
        }
        if (pos > size) {
            pos = size
        }
        super.add(pos, effect)
        effect.checkMarker = id
        id++
    }

    fun removeId(id: Int) {
        var i = 0
        while (i < size) {
            val effect = get(i)
            if (effect.checkMarker == id) {
                removeAt(i)
                i--
            }
            i++
        }
    }
}